coalib/core/DependencyTracker.py

```diff
@@ -0,0 +1,166 @@
+        """
+        if dependency not in self._dependency_dict:
+            self._dependency_dict[dependency] = set()
+
+        self._dependency_dict[dependency].add(dependant)
+
+    def resolve(self, dependency):
+        """
+        Resolves all dependency-relations from the given dependency, and frees
+        and returns dependants with no more dependencies. If the given
+        dependency is itself a dependant, all those relations are also removed.
+
+        >>> tracker = DependencyTracker()
+        >>> tracker.add(0, 1)
+        >>> tracker.add(0, 2)
+        >>> tracker.add(2, 3)
+        >>> tracker.resolve(0)
+        {1, 2}
+        >>> tracker.resolve(2)
+        {3}
+        >>> tracker.resolve(2)
+        set()
+
+        :param dependency:
+            The dependency.
+        :return:
+            Returns a set of dependants whose dependencies were all resolved.
+        """
+        # Check if dependency has itself dependencies which aren't resolved,
+        # these need to be removed too. This operation does not free any
+        # dependencies.
+        dependencies_to_remove = []
+        for tracked_dependency, dependants in self._dependency_dict.items():
+            if dependency in dependants:
+                dependants.remove(dependency)
+
+                # If dependants set is now empty, schedule dependency for
+                # removal from dependency_dict.
+                if not dependants:
+                    dependencies_to_remove.append(tracked_dependency)
+
+        for tracked_dependency in dependencies_to_remove:
+            del self._dependency_dict[tracked_dependency]
+
+        # Now free dependants which do depend on the given dependency.
+        possible_freed_dependants = self._dependency_dict.pop(
+            dependency, set())
+        non_free_dependants = set()
+
+        for possible_freed_dependant in possible_freed_dependants:
+            # Check if all dependencies of dependants from above are satisfied.
+            # If so, there are no more dependencies for dependant. Thus it's
+            # resolved.
+            for dependants in self._dependency_dict.values():
+                if possible_freed_dependant in dependants:
+                    non_free_dependants.add(possible_freed_dependant)
+                    break
+
+        # Remaining dependents are officially resolved.
+        return possible_freed_dependants - non_free_dependants
+
+    def check_circular_dependencies(self):
+        """
+        Checks whether there are circular dependency conflicts.
+
+        >>> tracker = DependencyTracker()
+        >>> tracker.add(0, 1)
+        >>> tracker.add(1, 0)
+        >>> tracker.check_circular_dependencies()
+        Traceback (most recent call last):
+         ...
+        coalib.core.CircularDependencyError.CircularDependencyError: ...
+
+        :raises CircularDependencyError:
+            Raised on circular dependency conflicts.
+        """
+        traverse_graph(
```

Comment:
circular dependencies can't be resolved, so yes, it should never happen :) The core would deadlock^^